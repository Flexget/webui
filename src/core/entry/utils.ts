import { RawEntry, FieldNames, MappingType, GettersType, BaseEntry } from './types';
import { MovieEntry, movieFieldList, RawMovieEntry, MovieFieldNames } from './fields/movies';
import { SeriesEntry, seriesFieldList, SeriesFieldNames, RawSeriesEntry } from './fields/series';
import {
  EpisodeEntry,
  RawEpisodeEntry,
  episodesFieldList,
  EpisodeFieldNames,
} from './fields/episodes';

export const isMovie = (entry: RawEntry): entry is RawMovieEntry => !!entry.movieName;
export const isSeries = (entry: RawEntry): entry is RawSeriesEntry =>
  !!entry.seriesName && !entry.seriesSeason;
export const isEpisode = (entry: RawEntry): entry is RawEpisodeEntry & SeriesEntry =>
  !!(entry.seriesSeason && entry.seriesEpisode && entry.seriesName);

const getPropFn = <U>(mapping: MappingType<U>) => {
  const mapByField = mapping.reduce<Partial<Record<FieldNames<U>, string[]>>>(
    (res, item) =>
      Object.entries(item).reduce(
        (out, [k, v]) => ({
          ...out,
          [k]: [...(out[k] ?? []), v],
        }),
        res,
      ),
    {},
  );
  return (entry: U, field: FieldNames<U>) =>
    (mapByField[field] ?? ([] as string[])).reduce((res, val) => res ?? entry[val], undefined);
};

const makeGetters = <T>(names: Record<string, FieldNames<T>>, mapping: MappingType<T>) => {
  const getProp = getPropFn<T>(mapping);
  return (entry: T) =>
    Object.values(names).reduce(
      (obj, key: FieldNames<T>) => ({
        ...obj,
        get [key]() {
          return getProp(entry, key);
        },
      }),
      {} as GettersType<T>,
    );
};

const getMovieProps = makeGetters<RawMovieEntry>(MovieFieldNames, movieFieldList);
const getSeriesProps = makeGetters<RawSeriesEntry>(SeriesFieldNames, seriesFieldList);
const getEpisodeProps = makeGetters<RawEpisodeEntry>(EpisodeFieldNames, episodesFieldList);

export const toMovieEntry = (entry: RawMovieEntry): MovieEntry => {
  return {
    ...entry,
    ...getMovieProps(entry),
  };
};

export const toSeriesEntry = (entry: RawSeriesEntry): SeriesEntry => ({
  ...entry,
  ...getSeriesProps(entry),
});

export const toEpisodeEntry = (entry: RawEpisodeEntry): EpisodeEntry => ({
  ...entry,
  ...getEpisodeProps(entry),
});

export const toDefaultEntry = (entry: BaseEntry): BaseEntry => ({
  ...entry,
});
