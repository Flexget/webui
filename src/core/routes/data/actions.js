const PREFIX = '@flexget/routes/';
export const ADD_ROUTE = `${PREFIX}ADD_ROUTE`;
