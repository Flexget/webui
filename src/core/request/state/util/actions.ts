import { StatusError } from 'utils/fetch';
import {
  AsyncKeys,
  Action,
  InProgressMeta,
  SuccessMeta,
  ErrorMeta,
  RequestState,
  UnknownAction,
  AsyncAction,
  RequestMeta,
  ActionFromState,
  InProgressAction,
  SuccessAction,
  ErrorAction,
} from './types';

interface ActionFunc {
  <T extends string>(type: T): Action<T>;
  <T extends string, P, U = {}>(type: T, payload: P, meta?: U): Action<T, P, U>;
}

export const action: ActionFunc = (type, payload = undefined, meta = {}) => ({
  type,
  meta,
  ...(payload && { payload }),
});

export const withMeta = <T extends string, P, U>(act: Action<T, P>, meta: U): Action<T, P, U> => ({
  ...act,
  meta: {
    ...(act.meta || {}),
    ...meta,
  },
});

// export const constantsCreator = (prefix: string) => (type: string) => ({
// [AsyncKeys.Request]: string(`${prefix}/${type}_REQUESTED`),
// [AsyncKeys.Success]: string(`${prefix}/${type}_SUCCESS`),
// [AsyncKeys.Failure]: string(`${prefix}/${type}_FAILURE`),
// });

interface RequestFunc {
  <T extends string, P>(type: T, payload: P, id?: string): Action<T, P, InProgressMeta>;
  <T extends string>(type: T): Action<T, undefined, InProgressMeta>;
}

interface SuccessFunc {
  <T extends string, P>(type: T, payload: P, messages?: string, id?: string): Action<
    T,
    P,
    SuccessMeta
  >;
  <T extends string>(type: T): Action<T, undefined, SuccessMeta>;
}

interface ErrorFunc {
  <T extends string, P>(type: T, err: StatusError, payload?: P, id?: string): Action<
    T,
    P,
    ErrorMeta
  >;
  <T extends string>(type: T, err: StatusError): Action<T, undefined, ErrorMeta>;
}

export const requestAction: RequestFunc = (type, payload = undefined, id = undefined) =>
  action<typeof type, typeof payload, InProgressMeta>(type, payload, {
    ...(id && { id }),
    state: RequestState.InProgress,
  });

export const successAction: SuccessFunc = (
  type,
  payload = undefined,
  message = undefined,
  id = undefined,
) =>
  action<typeof type, typeof payload, SuccessMeta>(type, payload, {
    ...(id && { id }),
    ...(message && { message }),
    state: RequestState.Success,
  });

export const errorAction: ErrorFunc = (type, err, payload = undefined, id = undefined) =>
  action<typeof type, typeof payload, ErrorMeta>(type, payload, {
    message: err.message,
    statusCode: err.status || 0,
    ...(id && { id }),
    state: RequestState.Error,
  });

export const isRequestAction = (act: UnknownAction): act is AsyncAction =>
  !!(act.meta && (act.meta as RequestMeta).state);

export const asyncAction = {
  [AsyncKeys.Request]: requestAction,
  [AsyncKeys.Success]: successAction,
  [AsyncKeys.Failure]: errorAction,
};

const isInStateHelper = <
  ActionUnion extends UnknownAction,
  T extends string,
  U extends RequestState
>(
  state: U,
  act: ActionUnion,
  type?: T,
): act is ActionFromState<ActionUnion, typeof type extends undefined ? string : T, U> =>
  isRequestAction(act) && act.meta.state === state && (type ? act.type === type : true);

export const isInState = {
  [RequestState.InProgress]: <
    ActionUnion extends UnknownAction = UnknownAction,
    ActionType extends string = string
  >(
    act: ActionUnion,
    type?: ActionType,
  ): act is InProgressAction<ActionUnion, ActionType> =>
    isInStateHelper(RequestState.InProgress, act, type),
  [RequestState.Success]: <
    ActionUnion extends UnknownAction = UnknownAction,
    ActionType extends string = string
  >(
    act: ActionUnion,
    type?: ActionType,
  ): act is SuccessAction<ActionUnion, ActionType> =>
    isInStateHelper(RequestState.Success, act, type),
  [RequestState.Error]: <
    ActionUnion extends UnknownAction = UnknownAction,
    ActionType extends string = string
  >(
    act: ActionUnion,
    type?: ActionType,
  ): act is ErrorAction<ActionUnion, ActionType> => isInStateHelper(RequestState.Error, act, type),
};

export const requesting = <T extends string>(type: T) => (
  act: UnknownAction,
): act is InProgressAction<T> => isInState.inProgress(act, type);
