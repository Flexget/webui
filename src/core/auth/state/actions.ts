import { asyncAction, ActionsUnion, RequestsOfType as ROT } from 'core/request/state/util';

import { StatusError } from 'utils/fetch';

export const enum Constants {
  LOGIN = '@flexget/auth/LOGIN',
  LOGOUT = '@flexget/auth/LOGOUT',
}

const asyncActions = {
  login: {
    request: (username: string, password: string) =>
      asyncAction.request(Constants.LOGIN, { username, password }),
    success: () => asyncAction.success(Constants.LOGIN),
    failure: (err: StatusError) => asyncAction.failure(Constants.LOGIN, err),
  },
  logout: {
    request: () => asyncAction.request(Constants.LOGOUT),
    success: () => asyncAction.success(Constants.LOGOUT),
    failure: (err: StatusError) => asyncAction.failure(Constants.LOGOUT, err),
  },
};

export type ActionTypes = ActionsUnion<typeof asyncActions>;
export type RequestsOfType<T extends Constants> = ROT<ActionTypes, T>;

export default asyncActions;
