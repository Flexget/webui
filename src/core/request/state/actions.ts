import { action, ActionsUnion } from './util';

export const enum Constants {
  CLEAR_ALL_REQUESTS = '@flexget/request/CLEAR_ALL_REQUESTS',
  CLEAR_REQUEST = '@flexget/request/CLEAR_REQUEST',
}

const actions = {
  clearAll: () => action(Constants.CLEAR_ALL_REQUESTS),
  clear: (requestActions: string[]) => action(Constants.CLEAR_REQUEST, { requestActions }),
};

export type ActionTypes = ActionsUnion<typeof actions>;

export default actions;
