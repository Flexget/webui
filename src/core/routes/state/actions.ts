import { RouteConfig } from 'core/registry/types';
import { action, ActionsUnion } from 'utils/actions';

export const enum Constants {
  ADD_ROUTE = '@flexget/routes/ADD_ROUTE',
}

const actions = {
  addRoute: (cfg: RouteConfig) => action(Constants.ADD_ROUTE, cfg),
};

export type ActionTypes = ActionsUnion<typeof actions>;

export default actions;
