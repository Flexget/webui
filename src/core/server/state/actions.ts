import statusActions from 'core/status/state/actions';
import { RequestsOfType as ROT } from 'core/status/state/types';
import { action, ActionsUnion } from 'utils/actions';
import { StatusError } from 'utils/fetch';

export const enum Constants {
  SERVER_OPERATION = '@flexget/server/SERVER_OPERATION',
}

export const enum Operation {
  Reload = 'reload',
  Shutdown = 'shutdown',
}

const actions = {
  serverOperation: {
    request: (operation: Operation) =>
      statusActions.load(Constants.SERVER_OPERATION, { operation }),
    success: () => action(Constants.SERVER_OPERATION),
    failure: (err: StatusError) => statusActions.error(Constants.SERVER_OPERATION, err),
  },
};

export type ActionTypes = ActionsUnion<typeof actions>;
export type RequestsOfType<T extends Constants> = ROT<ActionTypes, T>;

export default actions;
