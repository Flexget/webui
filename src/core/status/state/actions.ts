import { action, ActionsUnion, Action, Meta } from 'utils/actions';
import { StatusError } from 'utils/fetch';

export const enum Constants {
  LOADING_STATUS = 'LOADING_STATUS',
  ERROR_STATUS = 'ERROR_STATUS',
  CLOSE_STATUS = 'CLOSE_STATUS',
  INFO_STATUS = 'INFO_STATUS',
}

export function clearStatus() {
  return {
    type: Constants.CLOSE_STATUS,
  };
}

export interface TypeMeta<T extends string> extends Meta<string> {
  readonly type: T;
}

function load<T extends string, P>(
  type: T,
  payload: P,
): Action<Constants.LOADING_STATUS, P, TypeMeta<T>>;
function load<T extends string>(type: T): Action<Constants.LOADING_STATUS, undefined, TypeMeta<T>>;
function load<T extends string>(type: T, payload = undefined) {
  return action<
    Constants.LOADING_STATUS,
    typeof payload extends infer P ? P : undefined,
    TypeMeta<T>
  >(Constants.LOADING_STATUS, payload, {
    type,
  });
}

const actions = {
  load,
  error: <T extends string>(type: T, err: StatusError) =>
    action(Constants.ERROR_STATUS, {
      message: err.message,
      statusCode: err.status,
      type,
    }),
  clear: () => action(Constants.CLOSE_STATUS),
  info: <T extends string>(type: T, message: string) =>
    action(Constants.INFO_STATUS, { type, message }),
};

export type ActionTypes = ActionsUnion<typeof actions>;

export default actions;
