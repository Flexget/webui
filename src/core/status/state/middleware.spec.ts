import configureMockStore from 'redux-mock-store';
import { action, withMeta } from 'utils/actions';
import statusMiddleware from './middleware';
import actions from './actions';

const ACTION = 'ACTION';
const mockStore = configureMockStore([statusMiddleware]);
const store = mockStore({});

describe('core/status/data/middleware', () => {
  afterEach(() => store.clearActions());

  it('should dispatch INFO_STATUS and the original action if message is set', () => {
    const message = 'A message';
    const act = withMeta(action(ACTION), { message });
    store.dispatch(act);
    expect(store.getActions()).toEqual([actions.info(ACTION, message), act]);
  });
});
