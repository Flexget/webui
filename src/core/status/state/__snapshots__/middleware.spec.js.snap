// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`core/status/data/middleware should dispatch ERROR_STATUS if error and not ignore 1`] = `
Array [
  Object {
    "error": true,
    "payload": Object {
      "message": "Unauthorized",
      "statusCode": 401,
      "type": "ACTION",
    },
    "type": "ERROR_STATUS",
  },
]
`;

exports[`core/status/data/middleware should dispatch INFO_STATUS and the original action if message is set 1`] = `
Array [
  Object {
    "payload": Object {
      "message": "A message",
      "type": "ACTION",
    },
    "type": "INFO_STATUS",
  },
  Object {
    "error": false,
    "meta": Object {
      "message": "A message",
    },
    "payload": Object {},
    "type": "ACTION",
  },
]
`;

exports[`core/status/data/middleware should dispatch LOADING_STATUS if loading 1`] = `
Array [
  Object {
    "meta": Object {
      "type": "ACTION",
    },
    "payload": Object {},
    "type": "LOADING_STATUS",
  },
]
`;

exports[`core/status/data/middleware should not dispatch ERROR_STATUS if error and ignore 1`] = `
Array [
  Object {
    "error": true,
    "payload": Object {
      "message": "Unauthorized",
      "statusCode": 401,
      "type": "ACTION",
    },
    "type": "ERROR_STATUS",
  },
]
`;
